File: service/models/intervention.py

"""
Intervention 모델 (최소 버전)
"""
from dataclasses import dataclass
from datetime import datetime
from typing import Optional
from enum import Enum


class InterventionStatus(Enum):
    """개입 상태"""
    PENDING = "pending"
    SHOWN = "shown"
    INTERACTED = "interacted"
    DISMISSED = "dismissed"


class InterventionReason(Enum):
    """개입 이유"""
    NO_RECENT_RECORD = "no_recent_record"
    NEGATIVE_PATTERN = "negative_pattern"
    POSITIVE_REINFORCEMENT = "positive_reinforcement"


class MessageType(Enum):
    """메시지 유형 — 피드백 수집 대상 분류"""
    EMPATHY = "empathy"           # 부정 감정 패턴 공감
    ENCOURAGEMENT = "encouragement"  # 긍정 패턴 격려
    CHECKIN = "checkin"           # 장기 미기록 안부

REASON_TO_MESSAGE_TYPE: dict[str, str] = {
    "negative_pattern":      MessageType.EMPATHY.value,
    "negative_ratio":        MessageType.EMPATHY.value,
    "positive_reinforcement": MessageType.ENCOURAGEMENT.value,
    "no_recent_record":      MessageType.CHECKIN.value,
}


@dataclass
class Intervention:
    """
    개입 모델 (최소 버전)

    필수 필드만 포함. 나중에 필요하면 추가 가능.
    """
    # 필수 필드
    user_id: str
    reason: str
    message: str

    # 선택 필드
    id: Optional[str] = None
    status: str = InterventionStatus.PENDING.value
    message_type: Optional[str] = None
    created_at: Optional[datetime] = None

    def to_db_dict(self):
        """DB 저장용 딕셔너리 변환"""
        return {
            'user_id': self.user_id,
            'reason': self.reason,
            'message': self.message,
            'status': self.status,
            'message_type': self.message_type,
        }

    @classmethod
    def from_db_dict(cls, data: dict) -> 'Intervention':
        """DB 데이터에서 객체 생성"""
        return cls(
            id=data.get('id'),
            user_id=data['user_id'],
            reason=data['reason'],
            message=data['message'],
            status=data.get('status', 'pending'),
            message_type=data.get('message_type'),
            created_at=data.get('created_at')
        )
    
    def __repr__(self):
        return f"Intervention(id={self.id}, reason={self.reason}, status={self.status})"

File: lib/services/intervention.ts
import { getSupabaseBrowserClient } from "@/lib/supabase/client"

// ---------- Types ----------

export type Intervention = {
  id: string
  reason: string
  message: string
  status: "pending" | "shown" | "interacted" | "dismissed"
  message_type: "empathy" | "encouragement" | "checkin" | null
  created_at: string
}

// ---------- Functions ----------

export async function getLatestPendingIntervention(): Promise<Intervention | null> {
  const supabase = getSupabaseBrowserClient()

  const { data: { user } } = await supabase.auth.getUser()
  if (!user) return null

  const { data, error } = await supabase
    .from("interventions")
    .select("*")
    .eq("status", "pending")
    .eq("user_id", user.id)
    .order("created_at", { ascending: false })
    .limit(1)

  if (error) throw error
  if (!data || data.length === 0) return null
  return data[0] as Intervention
}

export async function markInterventionAsShown(id: string): Promise<void> {
  const supabase = getSupabaseBrowserClient()
  const { error } = await supabase
    .from("interventions")
    .update({ status: "shown" })
    .eq("id", id)

  if (error) throw error
}

export async function markInterventionAsInteracted(id: string): Promise<void> {
  const supabase = getSupabaseBrowserClient()
  const { error } = await supabase
    .from("interventions")
    .update({ status: "interacted" })
    .eq("id", id)

  if (error) throw error
}

export async function submitFeedback(
  interventionId: string,
  explicitScore: 2 | -2
): Promise<void> {
  const supabase = getSupabaseBrowserClient()
  const { data: { user } } = await supabase.auth.getUser()
  if (!user) return

  const { error } = await supabase
    .from("intervention_feedback")
    .insert({
      intervention_id: interventionId,
      user_id: user.id,
      explicit_score: explicitScore,
    })

  if (error) throw error
}
